src/main/cc/wfa/measurement/internal/duchy/protocol/share_shuffle/honest_majority_share_shuffle_utility.cc

```diff
@@ -126,7 +143,143 @@ absl::StatusOr<CompleteShufflePhaseResponse> CompleteShufflePhase(
+  // Shuffle the shares.
+  RETURN_IF_ERROR(SecureShuffleWithSeed(combined_sketch, shuffle_seed));
+
+  response.mutable_combined_sketch()->Add(combined_sketch.begin(),
+                                          combined_sketch.end());
+  *response.mutable_elapsed_cpu_duration() =
+      google::protobuf::util::TimeUtil::MillisecondsToDuration(
+          timer.ElapsedMillis());
+  return response;
+}
+
+absl::StatusOr<CompleteShufflePhaseResponse> CompleteReachOnlyShufflePhase(
+    const CompleteShufflePhaseRequest& request) {
+  StartedThreadCpuTimer timer;
+  CompleteShufflePhaseResponse response;
+
+  RETURN_IF_ERROR(VerifySketchParameters(request.sketch_params()));
+
+  // Verify that the ring modulus is a prime.
+  ASSIGN_OR_RETURN(bool isPrime,
+                   IsPrime(request.sketch_params().ring_modulus()));
+  if (!isPrime) {
+    return absl::InvalidArgumentError("The ring modulus must be a prime.");
+  }
+
+  if (request.sketch_shares().empty()) {
+    return absl::InvalidArgumentError("Sketch shares must not be empty.");
+  }
+
+  const int sketch_size = request.sketch_params().register_count();
+
+  // Combines the input shares.
+  std::vector<uint32_t> combined_sketch(sketch_size, 0);
+  for (int i = 0; i < request.sketch_shares().size(); i++) {
+    ASSIGN_OR_RETURN(
+        std::vector<uint32_t> share_vector,
+        GetShareVectorFromSketchShare(request.sketch_params(),
+                                      request.sketch_shares().Get(i)));
+    if (share_vector.size() != sketch_size) {
+      return absl::InvalidArgumentError(absl::Substitute(
+          "The $0-th sketch share has invalid size. Expect $1 but the "
+          "actual is $2.",
+          i, sketch_size, share_vector.size()));
+    }
+    for (int j = 0; j < sketch_size; j++) {
+      // It's guaranteed that (combined_sketch[j] + share_vector[j]) is
+      // not greater than 2^{32}-1.
+      combined_sketch[j] = (combined_sketch[j] + share_vector[j]) %
+                           request.sketch_params().ring_modulus();
+    }
+  }
+
+  ASSIGN_OR_RETURN(PrngSeed seed,
+                   GetPrngSeedFromString(request.common_random_seed()));
+  // Initializes the pseudo-random generator with the common random seed.
+  // The PRNG will generate random shares for the noise registers (if needed)
+  // and the seed that is used for shuffling.
+  ASSIGN_OR_RETURN(std::unique_ptr<UniformPseudorandomGenerator> prng,
+                   CreatePrngFromSeed(seed));
+
+  // Sample a vector r of random values in [1, modulus).
+  ASSIGN_OR_RETURN(std::vector<uint32_t> r,
+                   prng->GenerateNonZeroUniformRandomRange(
+                       sketch_size, request.sketch_params().ring_modulus()));
+
+  // Transform share of non-zero registers to share of a non-zero random value.
+  for (int j = 0; j < sketch_size; j++) {
+    combined_sketch[j] =
+        (combined_sketch[j] * r[j]) % request.sketch_params().ring_modulus();
```

Comment:
Maybe it is also worth having a CHECK here?  As an alternative we can use 64 bit arithmetic here e.g. change the type of 'r' to 'std::vector<uint64_t>'.